### ISSUE TITLE
We need to define all the required parameter widgets

### ISSUE THREAD

**Issue by timlinux** [inasafe/inasafe#1785]
# Problem

We need a well defined, well designed set of parameters for InaSAFE to use.
# Proposal
- [ ] @Jannes123: Audit all IFS and catalogue the Parameters needed by each one into the table below 
- [ ] @Jannes123 : Identify IF's that use parameters incorrectly e.g. having hardcoded attributes in the code (see for example ( https://github.com/AIFDR/inasafe/commit/84e2791ea96476745309f77f3cca61555775f1ad#diff-b83d63ac73b5a7c401e2864e7055d717 )
- [ ]  @lucernae - Create wireframes with clear use cases for each parameter widget
- [ ] @lucernae  - Move all parameters and widget code from InaSAFE repo (e.g. safe/common/resource_parameter.py)  into parameters repo (https://github.com/AIFDR/parameters) and use sentinels (test imports in try..except blocks) to determine if parameters should be made available or not. 
# Source of discussion
-  Spreadsheet files containing all the parameters exists in the IFs and what widget they associate.
  (https://drive.google.com/open?id=1o8xRZ_Sa3gkt0H8sLrXwJg4OyBKKqtMjQG8h7scRkW8&authuser=1)
- Moqups file to view possible widget implementations.
  (https://moqups.com/tim@kartoza.com/Zr0WXSV0)
- Feel free to comment and modify the spreadsheet and the moqups.
# CC

@Jannes123 
@lucernae 
@Charlotte-Morgan 
@akbargumbira 
@ismailsunni

**Comment by Jannes123** [comment]
I had to move table to excel as it wass getting big and messy in here. I will also paste doc in HipChat
![parameter_if_table](https://cloud.githubusercontent.com/assets/11268911/7135208/9753c834-e2a4-11e4-8d8e-3c7d733781f1.png)

**Comment by Jannes123** [comment]
@lucernae and @timlinux I think the

**Comment by timlinux** [comment]
Yeah that looks great

**note** I think all layer lookup stuff should really come from keywords of hazard / exposure layer rather
- and impact function should be able to just say "hazard layer, which is your inundated column"?

**Comment by timlinux** [comment]
BTW even better would be to make a shared google doc so that others can edit it etc.

**Comment by akbargumbira** [comment]
Agreed. It is listed on my google docs for phase 2 of IF refactoring

**Comment by akbargumbira** [comment]
Ah, just read the issue. I did ask @lucernae 2 weeks ago to document all the options in each IFs so that in phase 2 we can separate which options that could be taken from keywords (or naturally from it). Some of the options may have not be written to the current metadata system yet. So, I also asked @ismailsunni to look at this so that we can start working on it.

**Comment by akbargumbira** [comment]
The documents more or less list options like this (Excel would be better, with columns telling where the option should be taken: exposure layer, hazard layer, aggregation layer, IF)
e.g Flood Vector Building Impact QGIS. Options:
1. 'building_type_field', 'TYPE' # We should get it from exposure layer
2. 'affected_field', 'affected' # We should get it from hazard layer
3. 'affected_value', '1' # We should get this from hazard layer

**Comment by lucernae** [comment]
Should we differentiate on which tab the parameters that is currently being reviewed?
We should separate parameters that is used by the impact function and the post processor.

So perhaps we can have like 2 matrices, the first one is for the impact function to use and the other one is used by the post processor. but CMIIW, every post processor parameters for the IF looks the same for me, is there any case where it would be different?

It is also better to list the individual parameter in an IF so we can know what it is.

**Comment by akbargumbira** [comment]
Hi @lucernae, yes we should separate the parameters used by impact functions and the postprocessors (right now we tend to call it all as 'IF options'), but from now on let's call it as IF parameters. 

I propose to separate IF parameters for 3.2 as this:
1. IF Options
This option should focus on the behavior of the impact function itself. The example of this options are:
- X, Y, or displacement rate variables in ITB / Pager IF.
- Thresholds in Flood IF, etc

2. Options from Hazard and Exposure Keywords
The example for this is:
- For the Flood Polygon Building QGIS IF: Building Type Field (the attribute should be defined in the exposure keywords), Affected Field, Affected Value (these two attributes should be defined in hazard keywords)
For IF to be able to recognize the layers metadata, I will work on this https://github.com/AIFDR/inasafe/issues/1788 for 3.2.
We should also work with @ismailsunni and @borysiasty on this as they need to add some steps in keyword wizard to assign the layers those new keywords. @lucernae is making a document to list all the options in IF, and we'll decide where each option belongs to and we'll inform @ismailsunni and @borysiasty afterwards, is that ok?

3. Postprocessor
In the IF parameters, user should just be able to check if they want to include the postprocessor in the report. User shouldn't be able to update the value e.g age ratio. This means that all the postprocessor parameters will only be boolean parameters.
To implement this comprehensively, we should also work on this https://github.com/AIFDR/inasafe/issues/1726 (cc @cchristelis). The initial implementation would be:
- Removing the ability to update the postprocessor value (age ratio) in the IF parameters
- Adding 'Default' tab in InaSAFE Option for user to edit all the default values (gender and age ratio at the moment)
- Changing the UI and UX of assigning aggregation keywords in Keywords Wizard like the explanation in the issue (see the images in the proposed solution of isssue #1726)

After #1726 sorted out, the behavior of taking the value of the gender or age ratio would be:
- If user works in aoi mode, then take it from the settings in default tab
- If user works in aggregation mode, then take it from what's defined in aggregation layers (#1726)

4. Minimum Needs
I would just keep it like it is for 3.2

5. Hardcoded Option
We should get rid of this and categorise to one of those above options. The example of this is in flood_population_evacuation_polygon_hazard, it hardcodes the affected attribute (should belong to 2)

**Comment by Jannes123** [comment]
@akbargumbira and @lucernae   perhaps I don't understand your questions and comments above but I think definitely the postprocessors will not be integrated into the IF because the results can be obtained seperately.   Furthermore the postprocessors are separated in the postprocessors tab and should only be used by the postprocessor manager or related function.  @cchristelis please comment

**Comment by timlinux** [comment]
> I propose to separate IF parameters for 3.2 as this:
> # IF Options
> This option should focus on the behavior of the impact function itself. The example of this options are:
> - X, Y, or displacement rate variables in ITB / Pager IF.
> - Thresholds in Flood IF, etc

Agreed. In some cases (e.g. pager) we should hard code the parameters and not present them to the user. I do not think most users will understand the significance of these parameters or how to modify them in a meaningful way. 

> # Options from Hazard and Exposure Keywords
> The example for this is:
> - For the Flood Polygon Building QGIS IF: Building Type Field (the attribute should be defined in the
> exposure keywords), Affected Field, Affected Value (these two attributes should be defined in hazard
> keywords)

Agreed - and part of the validate() method should be to check that there are needed keywords in the Hazard or Exposure layers. One thing that might be an issue (though I think we should ignore it for now) is the case where the user has one layer with multiple columns representing e.g. which polygons were inundated during different events. They might want to run multiple analysis, switching each time to a different attribute column representing inundated. In the current regime it is quite quick and easy to do it by just changing the attribute name in IF options dialog. Anyway l am just raising this as a potential hindrance, but I think in 99% of cases this won't apply and we should not cater for it right now.

> For IF to be able to recognize the layers metadata, I will work on this  
> https://github.com/AIFDR/inasafe/issues/1788 for 3.2.

Ok cool

> We should also work with @ismailsunni and @borysiasty on this as they need to add some steps in keyword wizard to assign the layers those new keywords. @lucernae is making a document to list all the options in IF, and we'll decide where each option belongs to and we'll inform @ismailsunni and @borysiasty afterwards, is that ok?

Yes please help on this @borysiasty 

> # Postprocessor
> In the IF parameters, user should just be able to check if they want to include the postprocessor in the report. User shouldn't be able to update the value e.g age ratio. This means that all the postprocessor parameters will only be boolean parameters.

Agreed

> After #1726 sorted out, the behavior of taking the value of the gender or age ratio would be:
> - If user works in aoi mode, then take it from the settings in default tab
> - If user works in aggregation mode, then take it from what's defined in aggregation layers (#1726)

Yes and the other behaviours specified in #1726 will apply 

> # Minimum Needs
> I would just keep it like it is for 3.2

Agreed

> # Hardcoded Option
> We should get rid of this and categorise to one of those above options. The example of this is in flood_population_evacuation_polygon_hazard, it hardcodes the affected attribute (should belong to Options from hazard / exposure layers)

Agreed. In some cases (e.g. pager) we should hard code options that are essentially constants used for the algorithm unless we thing there is real benefit in exposing them to the user.

> # Open Question
> - Are we going to include the postprocessing work in the IF itself? (i.e generating the postprocessing report in the IF which right now is not done in the IF). The existence of postprocessing parameters in the IF itself would suggest that by running the IF we can get the postprocessing reports

Yes this was my intention - the IF should be self contained and we should not need to do all that nasty orchestration stuff in the Dock or other places outside of the IF. This also makes sense when InaSAFE is used in other contexts like web and CLI.

**Comment by lucernae** [comment]
I am just commenting the spreadsheet file, because it seems you put the IF
options and postprocessor in the same matrix, and I suggest to separate it
to have a clear separation of concept.

**Comment by timlinux** [comment]
@lucernae @Jannes123 can you please put a link to the spreadsheet in the original ticket above. I am a bit confused now which is the **official** one as @akbargumbira mentioned he had also been working on a similar tabulation. Can you check with him and consolidate the effort if needed so that we have one canonical reference point?

Thanks

**Comment by Charlotte-Morgan** [comment]
Great discussion. I would like to consult the earth quake IF experts to get their thoughts on whether users should be encouraged / able to change the parameters in the ITB function.

**Comment by timlinux** [comment]
@Charlotte-Morgan yes good idea, thanks!

**Comment by lucernae** [comment]
@timlinux It is actually the same spreadsheet but on a different tab. I added the link to the spreadsheet and moqups files at the original ticket above.

**Comment by Charlotte-Morgan** [comment]
hi @timlinux @Jannes123 @lucernae @akbargumbira @ismailsunni 
I have consulted the EQ experts and this is my recommendation:
Parameters for EQ impact functions should be hard coded & users should not be able to change the values from established / published models; however users should have the option as to which of te models they run (ITB or Pager) or potentially the option to upload their own model.
In my mind this could work a little like the Minimum needs, the user could choose from a list of existing models or upload their own file with model parameters but unlike the Min Needs would not have the option to modify the published model parameter files. The models would also list the reference in a provenance field, just like min needs.

So in summary: all these options have to disappear & possibly be replaced with a select your model that then fills in the model provenance. The Min needs provenance is also needed .
![image](https://cloud.githubusercontent.com/assets/5743135/7224482/37bced02-e75d-11e4-9161-dbc22de8ec23.png)

Hope this is helpful. 
# cc
@davidr007

**Comment by Charlotte-Morgan** [comment]
should the minimum need provenance field be shown on the tab that says "parameters that can be modified for this IF" ??? I think this provenance box might belong elsewhere.

**Comment by timlinux** [comment]
That all sounds fine to me. In the interim for 3.1 can we simply remove all those options? Or shall we leave things as they are for 3.1 and work on this only in 3.2?

**Comment by timlinux** [comment]
For the other parts (user defined models etc.) lets have a bit of a think about it to see what will fit in nicely with our API too. I'd like to avoid making any more big architectural changes (other than those on our roadmap) unless really needed.

**Comment by Charlotte-Morgan** [comment]
my preference would be to remove the EQ parameters ASAP & do the other bits later

**Comment by timlinux** [comment]
Assigned the first part (blocking the 3.1release) to @Jannes123  - hard code params into the IF and remove them from the parms dialog. In next cycle we can figure out a way forward for user defined params.

**Comment by Charlotte-Morgan** [comment]
yes - user defined parameters for EQ IF are low low priority

**Comment by Jannes123** [comment]
EarthQuake building functions still presents low_threshold, medium_threshold, high_threshold in the Options tab for this release

**Comment by cchristelis** [comment]
The hard coded values, which where the release blocker, have been done and merged in #1793. As such the this is no longer a release blocker and the remaining work can be left for 3.2.

**Comment by ismailsunni** [comment]
I don't think it's compatible with our current InaSAFE architecture, we have put the parameter in the hazard layer's metadata and all is shown nicely in the wizard.
